src/main/java/com/pingcap/tikv/meta/TiDAGRequest.java

```diff
@@ -0,0 +1,436 @@
+    this.distinct = distinct;
+    return this;
+  }
+
+  public boolean isDistinct() {
+    return distinct;
+  }
+
+  /**
+   * add aggregate function to select query
+   *
+   * @param expr is a TiUnaryFunction expression.
+   * @return a SelectBuilder
+   */
+  public TiDAGRequest addAggregate(TiExpr expr) {
+    requireNonNull(expr, "aggregation expr is null");
+    aggregates.add(Pair.create(expr, expr.getType()));
+    return this;
+  }
+
+  public TiDAGRequest addAggregate(TiExpr expr, DataType targetType) {
+    requireNonNull(expr, "aggregation expr is null");
+    aggregates.add(Pair.create(expr, targetType));
+    return this;
+  }
+
+  public List<TiExpr> getAggregates() {
+    return aggregates.stream().map(p -> p.first).collect(Collectors.toList());
+  }
+
+  public List<Pair<TiExpr, DataType>> getAggregatePairs() {
+    return aggregates;
+  }
+
+  /**
+   * add a order by clause to select query.
+   *
+   * @param byItem is a TiByItem.
+   * @return a SelectBuilder
+   */
+  TiDAGRequest addOrderByItem(TiByItem byItem) {
+    orderByItems.add(requireNonNull(byItem, "byItem is null"));
+    return this;
+  }
+
+  List<TiByItem> getOrderByItems() {
+    return orderByItems;
+  }
+
+  /**
+   * add a group by clause to select query
+   *
+   * @param byItem is a TiByItem
+   * @return a SelectBuilder
+   */
+  public TiDAGRequest addGroupByItem(TiByItem byItem) {
+    groupByItems.add(requireNonNull(byItem, "byItem is null"));
+    return this;
+  }
+
+  public List<TiByItem> getGroupByItems() {
+    return groupByItems;
+  }
+
+  /**
+   * Field is not support in TiDB yet, for here we simply allow TiColumnRef instead of TiExpr like
+   * in SelectRequest proto
+   * <p>
+   * <p>This interface allows duplicate columns and it's user's responsibility to do dedup since we
+   * need to ensure exact order and items preserved during decoding
+   *
+   * @param column is column referred during selectReq
+   */
+  public TiDAGRequest addField(TiColumnRef column) {
+    fields.add(requireNonNull(column, "columnRef is null"));
+    return this;
+  }
+
+  public List<TiColumnRef> getFields() {
+    return fields;
+  }
+
+  /**
+   * set key range of scan
+   *
+   * @param ranges key range of scan
+   */
+  public TiDAGRequest addRanges(List<Coprocessor.KeyRange> ranges) {
+    keyRanges.addAll(requireNonNull(ranges, "KeyRange is null"));
+    return this;
+  }
+
+  public void resetRanges(List<Coprocessor.KeyRange> ranges) {
+    keyRanges.clear();
+    keyRanges.addAll(ranges);
+  }
+
+  public List<Coprocessor.KeyRange> getRanges() {
+    return keyRanges;
+  }
+
+  public TiDAGRequest addWhere(TiExpr where) {
+    this.where.add(requireNonNull(where, "where expr is null"));
+    return this;
+  }
+
+  /**
+   * Has aggregate expression.
+   *
+   * @return the boolean
+   */
+  public boolean hasAggregate() {
+    return !getAggregates().isEmpty();
+  }
+
+  /**
+   * Has group by expression.
+   *
+   * @return the boolean
```

Comment:
same as above comment.

Reply:
This is a function returns whether the `DAGRequest` has any group by statement. It indeed does not set anything but get whether there's any group by statement to the caller.

Reply:
Ok